### PR TITLE
feat: `swipe controls` override volume button behaviour

### LINKED ISSUE #279: feat: Unlock Cricbuzz Premium
### Application

Cricbuzz ( https://play.google.com/store/apps/details?id=com.cricbuzz.android )

### Issue

Unlock Cricbuzz Premium 

### Patch

To unlock the cricbuzz premium features. ( Removes ads + Unlock premium articles ) 

And also to add option to change theme as per system settings. 

### Motivation

Cricbuzz is one of the most popular sports app for Cricket. But they have started bombarding ads. Most of the time misleading and to some mild NSFW ads. Not only it ruins the whole experience but it makes the app less friendly for younger cricket enthusiasts.  

Their premium subscription isn't value for money and merely removes the ads and unlock few other articles to read. 

Also, though you can switch to dark mode in the app settings but there is not option of theme changing as per system settings. 

### Acknowledgements

- [X] I have searched the existing issues and this is a new and no duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I filled out all of the requested information in this issue properly.

### PR DESCRIPTION
**Pull request by shadow578**
closes https://github.com/revanced/revanced-patches/issues/279

needs to be merged together with https://github.com/revanced/revanced-integrations/pull/114

### CODE REVIEW

File: src/main/kotlin/app/revanced/extensions/Extensions.kt [hunk 1]
```diff
@@ -27,6 +29,30 @@ internal fun MutableMethodImplementation.injectHideCall(
...
 
+/**
+ * traverse the class hierarchy starting from the given root class
+ *
+ * @param targetClass the class to start traversing the class hierarchy from
+ * @param callback function that is called for every class in the hierarchy
+ */
+fun BytecodeData.traverseClassHierarchy(targetClass: MutableClass, callback: MutableClass.() -> Unit) {
+    callback(targetClass)
+    this.findClass(targetClass.superclass ?: "")?.resolve()?.let {
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/extensions/Extensions.kt:40]: Is resolving the proxy required? Resolve creates a mutable clone of ImmutableClass. You should rely on the immutable type as long as you do not need mutability.

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/revanced/extensions/Extensions.kt (hunk 1 above)
Comment (shadow578) [reply src/main/kotlin/app/revanced/extensions/Extensions.kt:40]: idea behind this is that, if you're traversing the class hierarchy, you'll probably also want to mutate the classes in some way. 

Changing the function to 
```kotlin
fun BytecodeData.traverseClassHierarchy(targetClass: ClassDef, callback: ClassDef.() -> Unit) {
    targetClass.callback()
    this.findClass(targetClass.superclass ?: "")?.immutableClass?.let {
        traverseClassHierarchy(it, callback)
    }
}
```
and resolving in the callback does not work, as the changes done to the mutable class (superclass is changed) are not reflected in `immutableClass`.

Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/extensions/Extensions.kt:40]: @Sculas, please check if this is really necessary.

**Comment by shadow578** [comment]
while trying to implement the changes requested, i found that it didn't actually work quite right (final flag on methods wasn't cleared correctly for whatever reason).

---

What the `traverseClassHierarchy` is used for is to remove the `final` modifier on all classes (and their methods) in all superclasses of a class, so that the injected class can overwrite them.
Maybe there is a better way to achive this, but i couldn't find one...

File: src/main/kotlin/app/revanced/extensions/Extensions.kt (hunk 1 above)
Comment (Sculas) [inline src/main/kotlin/app/revanced/extensions/Extensions.kt:40]: Finding a class named `""` is going to get you nowhere.
```suggestion
    this.findClass(targetClass.superclass ?: return)?.resolve()?.let {
```

**Review by Sculas — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/revanced/extensions/Extensions.kt (hunk 1 above)
Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/extensions/Extensions.kt:40]: @Sculas Please also review this

Comment (Sculas) [reply src/main/kotlin/app/revanced/extensions/Extensions.kt:40]: I'll look into it.

**Comment by oSumAtrIX** [comment]
Thanks

**Pull request merged**